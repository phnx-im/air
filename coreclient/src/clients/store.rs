// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use anyhow::bail;
use phnxtypes::messages::push_token::PushToken;

use super::{
    create_user::{
        AsRegisteredUserState, BasicUserData, PersistedUserState, PostRegistrationInitState,
        QsRegisteredUserState, UnfinalizedRegistrationState,
    },
    *,
};

/// WARNING: This enum is stored in sqlite as a blob. If any changes are made to
/// this enum, a new version in `StorableUserCreationState` must be created.
#[derive(Serialize, Deserialize)]
pub(crate) enum UserCreationState {
    BasicUserData(BasicUserData),
    InitialUserState(InitialUserState),
    PostRegistrationInitState(PostRegistrationInitState),
    UnfinalizedRegistrationState(UnfinalizedRegistrationState),
    AsRegisteredUserState(AsRegisteredUserState),
    QsRegisteredUserState(QsRegisteredUserState),
    FinalUserState(PersistedUserState),
}

impl UserCreationState {
    pub(super) fn client_id(&self) -> &AsClientId {
        match self {
            Self::BasicUserData(state) => state.client_id(),
            Self::InitialUserState(state) => state.client_id(),
            Self::PostRegistrationInitState(state) => state.client_id(),
            Self::UnfinalizedRegistrationState(state) => state.client_id(),
            Self::AsRegisteredUserState(state) => state.client_id(),
            Self::QsRegisteredUserState(state) => state.client_id(),
            Self::FinalUserState(state) => state.client_id(),
        }
    }

    pub(super) fn server_url(&self) -> &str {
        match self {
            Self::BasicUserData(state) => state.server_url(),
            Self::InitialUserState(state) => state.server_url(),
            Self::PostRegistrationInitState(state) => state.server_url(),
            Self::UnfinalizedRegistrationState(state) => state.server_url(),
            Self::AsRegisteredUserState(state) => state.server_url(),
            Self::QsRegisteredUserState(state) => state.server_url(),
            Self::FinalUserState(state) => state.server_url(),
        }
    }

    pub(super) fn new(
        client_db_connection: &Connection,
        phnx_db_connection: &Connection,
        as_client_id: AsClientId,
        server_url: impl ToString,
        password: &str,
        push_token: Option<PushToken>,
    ) -> Result<Self> {
        let client_record = ClientRecord::new(as_client_id.clone());
        client_record.store(phnx_db_connection)?;

        let basic_user_data = BasicUserData {
            as_client_id: as_client_id.clone(),
            server_url: server_url.to_string(),
            password: password.to_string(),
            push_token,
        };

        // Create user profile entry for own user.
        UserProfile::store_own_user_profile(
            client_db_connection,
            &mut StoreNotifier::noop(),
            as_client_id.user_name(),
            None,
            None,
        )?;

        let user_creation_state = UserCreationState::BasicUserData(basic_user_data);

        user_creation_state.store(client_db_connection)?;

        Ok(user_creation_state)
    }

    pub(super) async fn step(
        self,
        phnx_db_connection: SqliteConnection,
        client_db_connection: SqliteConnection,
        api_clients: &ApiClients,
    ) -> Result<Self> {
        // If we're already in the final state, there is nothing to do.
        if matches!(self, UserCreationState::FinalUserState(_)) {
            return Ok(self);
        }

        let new_state = match self {
            UserCreationState::BasicUserData(state) => Self::InitialUserState(
                state
                    .prepare_as_registration(client_db_connection.clone(), api_clients)
                    .await?,
            ),
            UserCreationState::InitialUserState(state) => {
                Self::PostRegistrationInitState(state.initiate_as_registration(api_clients).await?)
            }
            UserCreationState::PostRegistrationInitState(state) => {
                let connection = client_db_connection.lock().await;
                Self::UnfinalizedRegistrationState(state.process_server_response(&connection)?)
            }
            UserCreationState::UnfinalizedRegistrationState(state) => {
                Self::AsRegisteredUserState(state.finalize_as_registration(api_clients).await?)
            }
            UserCreationState::AsRegisteredUserState(state) => {
                Self::QsRegisteredUserState(state.register_with_qs(api_clients).await?)
            }
            UserCreationState::QsRegisteredUserState(state) => Self::FinalUserState(
                state
                    .upload_key_packages(client_db_connection.clone(), api_clients)
                    .await?,
            ),
            UserCreationState::FinalUserState(_) => self,
        };

        let client_db_connection = client_db_connection.lock().await;
        new_state.store(&client_db_connection)?;

        // If we just transitioned into the final state, we need to update the
        // client record.
        let phnx_db_connection = phnx_db_connection.lock().await;
        if let UserCreationState::FinalUserState(_) = new_state {
            let mut client_record = ClientRecord::load(&phnx_db_connection, new_state.client_id())?
                .ok_or(anyhow!("Client record not found"))?;
            client_record.finish();
            client_record.store(&phnx_db_connection)?;
        }

        Ok(new_state)
    }

    pub(super) fn final_state(self) -> Result<PersistedUserState> {
        if let UserCreationState::FinalUserState(state) = self {
            Ok(state)
        } else {
            bail!("Unexpected user creation state")
        }
    }

    /// A convenience function that performs the `step` function until the final state is reached.
    pub(super) async fn complete_user_creation(
        mut self,
        phnx_db_connection: SqliteConnection,
        client_db_connection: SqliteConnection,
        api_clients: &ApiClients,
    ) -> Result<PersistedUserState> {
        while !matches!(self, UserCreationState::FinalUserState(_)) {
            self = self
                .step(
                    phnx_db_connection.clone(),
                    client_db_connection.clone(),
                    api_clients,
                )
                .await?
        }

        self.final_state()
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum ClientRecordState {
    InProgress,
    Finished,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ClientRecord {
    pub as_client_id: AsClientId,
    pub client_record_state: ClientRecordState,
    pub created_at: DateTime<Utc>,
    pub is_default: bool,
}

impl ClientRecord {
    pub(super) fn new(as_client_id: AsClientId) -> Self {
        Self {
            as_client_id,
            client_record_state: ClientRecordState::InProgress,
            created_at: Utc::now(),
            is_default: false,
        }
    }

    pub(super) fn finish(&mut self) {
        self.client_record_state = ClientRecordState::Finished;
    }
}
