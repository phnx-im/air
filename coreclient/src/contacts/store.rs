// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use tls_codec::Serialize;

use crate::utils::persistance::{DataType, Persistable};

use super::*;

impl Persistable for Contact {
    type Key = UserName;
    type SecondaryKey = UserName;
    const DATA_TYPE: DataType = DataType::Contact;

    fn own_client_id_bytes(&self) -> Vec<u8> {
        self.own_client_id.tls_serialize_detached().unwrap()
    }

    fn rowid(&self) -> Option<i64> {
        self.rowid
    }

    fn key(&self) -> &Self::Key {
        &self.user_name
    }

    fn secondary_key(&self) -> &Self::SecondaryKey {
        &self.user_name
    }

    fn set_rowid(&mut self, rowid: i64) {
        self.rowid = Some(rowid);
    }
}
