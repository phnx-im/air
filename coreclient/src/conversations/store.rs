// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use std::collections::HashMap;

use openmls::prelude::GroupId;
use phnxbackend::auth_service::UserName;
use serde::{Deserialize, Serialize};

use crate::types::*;

use super::*;

#[derive(Default, Serialize, Deserialize)]
pub(crate) struct ConversationStore {
    #[serde(
        serialize_with = "crate::utils::serialize_hashmap",
        deserialize_with = "crate::utils::deserialize_hashmap"
    )]
    conversations: HashMap<Uuid, Conversation>,
    messages: HashMap<Uuid, HashMap<Uuid, ConversationMessage>>,
}

impl ConversationStore {
    pub(crate) fn conversation_by_group_id(&self, group_id: &GroupId) -> Option<&Conversation> {
        self.conversations
            .values()
            .find(|conversation| conversation.group_id.as_group_id() == *group_id)
    }

    pub(crate) fn create_connection_conversation(
        &mut self,
        group_id: GroupId,
        user_name: UserName,
        attributes: ConversationAttributes,
    ) -> Uuid {
        // To keep things simple and to make sure that conversation ids are the
        // same across users, we derive the conversation id from the group id.
        let uuid_bytes = UuidBytes::from_group_id(&group_id);
        let conversation = Conversation {
            id: uuid_bytes.clone(),
            group_id: group_id.into(),
            status: ConversationStatus::Active,
            conversation_type: ConversationType::UnconfirmedConnection(user_name.to_string()),
            last_used: Timestamp::now().as_u64(),
            attributes,
        };
        self.conversations
            .insert(uuid_bytes.as_uuid().clone(), conversation);
        uuid_bytes.as_uuid()
    }

    pub(crate) fn create_group_conversation(
        &mut self,
        group_id: GroupId,
        attributes: ConversationAttributes,
    ) -> Uuid {
        let uuid_bytes = UuidBytes::from_group_id(&group_id);
        let conversation_id = uuid_bytes.as_uuid();
        let conversation = Conversation {
            id: uuid_bytes.clone(),
            group_id: group_id.into(),
            status: ConversationStatus::Active,
            conversation_type: ConversationType::Group,
            last_used: Timestamp::now().as_u64(),
            attributes,
        };
        self.conversations
            .insert(conversation_id.clone(), conversation);
        conversation_id
    }

    pub(crate) fn confirm_connection_conversation(&mut self, conversation_id: &Uuid) {
        if let Some(conversation) = self.conversations.get_mut(conversation_id) {
            if let ConversationType::UnconfirmedConnection(user_name) =
                conversation.conversation_type.clone()
            {
                conversation.conversation_type = ConversationType::Connection(user_name);
            }
        }
    }

    pub(crate) fn conversations(&self) -> Vec<Conversation> {
        let mut conversations: Vec<Conversation> = self.conversations.values().cloned().collect();
        conversations.sort_by(|a, b| a.last_used.cmp(&b.last_used));
        conversations
    }

    pub(crate) fn conversation(&self, conversation_id: Uuid) -> Option<&Conversation> {
        self.conversations.get(&conversation_id)
    }

    pub(crate) fn set_inactive(&mut self, conversation_id: Uuid, past_members: &[String]) {
        self.conversations
            .get_mut(&conversation_id)
            .map(|conversation| {
                conversation.status = ConversationStatus::Inactive(InactiveConversation {
                    past_members: past_members.iter().map(|m| m.to_owned()).collect(),
                })
            });
    }

    pub(crate) fn messages(
        &self,
        conversation_id: Uuid,
        last_n: usize,
    ) -> Vec<ConversationMessage> {
        match self.messages.get(&conversation_id) {
            Some(messages) => {
                let mut messages: Vec<ConversationMessage> = messages
                    .iter()
                    .map(|(_uuid, message)| message.clone())
                    .collect();
                messages.sort_by(|a, b| a.timestamp.cmp(&b.timestamp));
                if last_n >= messages.len() {
                    messages
                } else {
                    let (_left, right) = messages.split_at(messages.len() - last_n);
                    right.to_vec()
                }
            }
            None => {
                vec![]
            }
        }
    }

    pub(crate) fn store_message(
        &mut self,
        conversation_id: Uuid,
        message: ConversationMessage,
    ) -> Result<(), ConversationStoreError> {
        let message_id = message.id.clone();
        match self.conversations.get(&conversation_id) {
            Some(_conversation_data) => {
                match self.messages.get_mut(&conversation_id) {
                    Some(conversation_messages) => {
                        conversation_messages.insert(message_id.as_uuid(), message);
                    }
                    None => {
                        let mut conversation_messages = HashMap::new();
                        conversation_messages.insert(message_id.as_uuid(), message);
                        self.messages.insert(conversation_id, conversation_messages);
                    }
                }
                Ok(())
            }
            None => Err(ConversationStoreError::UnknownConversation),
        }
    }
}
