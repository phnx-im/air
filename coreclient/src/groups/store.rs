// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use turbosql::{execute, select, Turbosql};

use super::*;

impl ClientGroup {
    /// Writes the new group to the database. If a group with the same group id
    /// already exists, it deletes the old group.
    pub(crate) fn new(inner_group: InnerClientGroup) -> Result<Self, turbosql::Error> {
        let client_id_bytes = inner_group
            .client_by_index(inner_group.mls_group().own_leaf_index().usize())
            .unwrap()
            .tls_serialize_detached()
            .unwrap();
        let group_id_bytes = inner_group.group_id().as_slice().to_vec();
        // Check if a group with this ID already exists.
        if let Ok(old_group) = select!(ClientGroup "WHERE client_id = " client_id_bytes " AND group_id = " group_id_bytes)
        {
            // If it exists, delete it from the DB.
            execute!("DELETE FROM clientgroup WHERE rowid = " old_group.rowid.unwrap())?;
        }
        // Insert the new group into the DB.
        let group = Self {
            rowid: None,
            client_id: Some(client_id_bytes),
            group_id: Some(group_id_bytes),
            inner_client_group: Some(inner_group),
        };
        group.insert()?;
        Ok(group)
    }

    pub(crate) fn load(
        group_id: &GroupId,
        client_id: &AsClientId,
    ) -> Result<Self, turbosql::Error> {
        let client_id_bytes = client_id.tls_serialize_detached().unwrap();
        let group_id_bytes = group_id.as_slice();
        let group = select!(ClientGroup "WHERE client_id = " client_id_bytes " AND group_id = " group_id_bytes)?;
        if group.inner_client_group.is_none() {
            return Err(turbosql::Error::OtherError("Corrupted group in DB."));
        }
        Ok(group)
    }

    pub(crate) fn persist(&self) -> Result<(), turbosql::Error> {
        <Self as Turbosql>::update(self)?;
        Ok(())
    }

    pub(crate) fn _purge(&self) -> Result<(), turbosql::Error> {
        execute!("DELETE FROM clientgroup WHERE rowid = " self.rowid.unwrap())?;
        Ok(())
    }
}
