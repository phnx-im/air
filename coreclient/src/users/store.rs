// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use anyhow::bail;

use super::{
    create_user::{
        AsRegisteredUserState, PersistedUserState, PostRegistrationInitState,
        QsRegisteredUserState, UnfinalizedRegistrationState,
    },
    *,
};

#[derive(Serialize, Deserialize)]
pub(super) enum UserCreationState {
    InitialUserState(InitialUserState),
    PostRegistrationInitState(PostRegistrationInitState),
    UnfinalizedRegistrationState(UnfinalizedRegistrationState),
    AsRegisteredUserState(AsRegisteredUserState),
    QsRegisteredUserState(QsRegisteredUserState),
    FinalUserState(PersistedUserState),
}

impl UserCreationState {
    fn client_id(&self) -> &AsClientId {
        match self {
            Self::InitialUserState(state) => state.client_id(),
            Self::PostRegistrationInitState(state) => state.client_id(),
            Self::UnfinalizedRegistrationState(state) => state.client_id(),
            Self::AsRegisteredUserState(state) => state.client_id(),
            Self::QsRegisteredUserState(state) => state.client_id(),
            Self::FinalUserState(state) => state.client_id(),
        }
    }

    pub(super) fn server_url(&self) -> &str {
        match self {
            Self::InitialUserState(state) => state.server_url(),
            Self::PostRegistrationInitState(state) => state.server_url(),
            Self::UnfinalizedRegistrationState(state) => state.server_url(),
            Self::AsRegisteredUserState(state) => state.server_url(),
            Self::QsRegisteredUserState(state) => state.server_url(),
            Self::FinalUserState(state) => state.server_url(),
        }
    }
}

impl InitialUserState {
    pub(super) fn persist(self, connection: &Connection) -> Result<Self> {
        let user_data = PersistableUserData::from_connection_and_payload(
            connection,
            UserCreationState::InitialUserState(self),
        );
        user_data.persist()?;
        if let UserCreationState::InitialUserState(state) = user_data.into_payload() {
            Ok(state)
        } else {
            bail!("Unexpected user creation state")
        }
    }
}

impl PostRegistrationInitState {
    pub(super) fn persist(self, connection: &Connection) -> Result<Self> {
        let user_data = PersistableUserData::from_connection_and_payload(
            connection,
            UserCreationState::PostRegistrationInitState(self),
        );
        user_data.persist()?;
        if let UserCreationState::PostRegistrationInitState(state) = user_data.into_payload() {
            Ok(state)
        } else {
            bail!("Unexpected user creation state")
        }
    }
}

impl UnfinalizedRegistrationState {
    pub(super) fn persist(self, connection: &Connection) -> Result<Self> {
        let user_data = PersistableUserData::from_connection_and_payload(
            connection,
            UserCreationState::UnfinalizedRegistrationState(self),
        );
        user_data.persist()?;
        if let UserCreationState::UnfinalizedRegistrationState(state) = user_data.into_payload() {
            Ok(state)
        } else {
            bail!("Unexpected user creation state")
        }
    }
}

impl AsRegisteredUserState {
    pub(super) fn persist(self, connection: &Connection) -> Result<Self> {
        let user_data = PersistableUserData::from_connection_and_payload(
            connection,
            UserCreationState::AsRegisteredUserState(self),
        );
        user_data.persist()?;
        if let UserCreationState::AsRegisteredUserState(state) = user_data.into_payload() {
            Ok(state)
        } else {
            bail!("Unexpected user creation state")
        }
    }
}

impl QsRegisteredUserState {
    pub(super) fn persist(self, connection: &Connection) -> Result<Self> {
        let user_data = PersistableUserData::from_connection_and_payload(
            connection,
            UserCreationState::QsRegisteredUserState(self),
        );
        user_data.persist()?;
        if let UserCreationState::QsRegisteredUserState(state) = user_data.into_payload() {
            Ok(state)
        } else {
            bail!("Unexpected user creation state")
        }
    }
}

impl PersistedUserState {
    pub(super) fn persist(self, connection: &Connection) -> Result<Self> {
        let user_data = PersistableUserData::from_connection_and_payload(
            connection,
            UserCreationState::FinalUserState(self),
        );
        user_data.persist()?;
        if let UserCreationState::FinalUserState(state) = user_data.into_payload() {
            Ok(state)
        } else {
            bail!("Unexpected user creation state")
        }
    }
}

pub(super) struct PersistableUserData<'a> {
    connection: &'a Connection,
    payload: UserCreationState,
}

impl PersistableUserData<'_> {
    pub(super) fn into_payload(self) -> UserCreationState {
        self.payload
    }

    pub(super) fn server_url(&self) -> &str {
        self.payload.server_url()
    }
}

impl<'a> Persistable<'a> for PersistableUserData<'a> {
    type Key = AsClientId;

    type SecondaryKey = AsClientId;

    type Payload = UserCreationState;

    const DATA_TYPE: DataType = DataType::ClientData;

    fn key(&self) -> &Self::Key {
        &self.payload.client_id()
    }

    fn secondary_key(&self) -> &Self::SecondaryKey {
        &self.payload.client_id()
    }

    fn connection(&self) -> &Connection {
        self.connection
    }

    fn payload(&self) -> &Self::Payload {
        &self.payload
    }

    fn from_connection_and_payload(conn: &'a Connection, payload: Self::Payload) -> Self {
        Self {
            connection: conn,
            payload,
        }
    }
}
