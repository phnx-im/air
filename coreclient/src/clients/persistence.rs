// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use aircommon::{
    codec::AirCodec,
    identifiers::{Fqdn, UserId},
};
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::{
    Database, Decode, Encode, Sqlite, SqliteExecutor, Type, encode::IsNull, error::BoxDynError,
    query, query_as, query_scalar, sqlite::SqliteTypeInfo,
};
use uuid::Uuid;

use crate::{clients::store::EncryptedDek, utils::persistence::open_air_db};

use super::store::{ClientRecord, ClientRecordState, UserCreationState};

// When adding a variant to this enum, the new variant must be called
// `CurrentVersion` and the current version must be renamed to `VX`, where `X`
// is the next version number. The content type of the old `CurrentVersion` must
// be renamed and otherwise preserved to ensure backwards compatibility.
#[derive(Serialize, Deserialize)]
enum StorableUserCreationState {
    CurrentVersion(UserCreationState),
}

// Only change this enum in tandem with its non-Ref variant.
#[derive(Serialize)]
enum StorableUserCreationStateRef<'a> {
    CurrentVersion(&'a UserCreationState),
}

impl Type<Sqlite> for UserCreationState {
    fn type_info() -> SqliteTypeInfo {
        <Vec<u8> as Type<Sqlite>>::type_info()
    }
}

impl<'q> Encode<'q, Sqlite> for UserCreationState {
    fn encode_by_ref(
        &self,
        buf: &mut <Sqlite as Database>::ArgumentBuffer<'q>,
    ) -> Result<IsNull, BoxDynError> {
        let state = StorableUserCreationStateRef::CurrentVersion(self);
        let bytes = AirCodec::to_vec(&state)?;
        Encode::<Sqlite>::encode(bytes, buf)
    }
}

impl<'r> Decode<'r, Sqlite> for UserCreationState {
    fn decode(value: <Sqlite as Database>::ValueRef<'r>) -> Result<Self, BoxDynError> {
        let bytes: &[u8] = Decode::<Sqlite>::decode(value)?;
        let state = AirCodec::from_slice(bytes)?;
        match state {
            StorableUserCreationState::CurrentVersion(state) => Ok(state),
        }
    }
}

impl UserCreationState {
    pub(super) async fn load(
        executor: impl SqliteExecutor<'_>,
        user_id: &UserId,
    ) -> sqlx::Result<Option<Self>> {
        let uuid = user_id.uuid();
        let domain = user_id.domain();
        query_scalar!(
            r#"SELECT state AS "state: _"
            FROM user_creation_state WHERE user_uuid = ? AND user_domain = ?"#,
            uuid,
            domain
        )
        .fetch_optional(executor)
        .await
    }

    pub(super) async fn store(&self, executor: impl SqliteExecutor<'_>) -> sqlx::Result<()> {
        let user_id = self.user_id();
        let uuid = user_id.uuid();
        let domain = user_id.domain();
        query!(
            "INSERT OR REPLACE INTO user_creation_state
                (user_uuid, user_domain, state)
            VALUES (?, ?, ?)",
            uuid,
            domain,
            self
        )
        .execute(executor)
        .await?;
        Ok(())
    }
}

impl Type<Sqlite> for ClientRecordState {
    fn type_info() -> <Sqlite as Database>::TypeInfo {
        <&str as Type<Sqlite>>::type_info()
    }
}

impl<'q> Encode<'q, Sqlite> for ClientRecordState {
    fn encode_by_ref(
        &self,
        buf: &mut <Sqlite as Database>::ArgumentBuffer<'q>,
    ) -> Result<IsNull, BoxDynError> {
        Encode::<Sqlite>::encode(self.as_str(), buf)
    }
}

#[derive(Debug, thiserror::Error)]
#[error("Invalid ClientRecordState: {state}")]
struct InvalidClientRecordState {
    state: String,
}

impl<'r> Decode<'r, Sqlite> for ClientRecordState {
    fn decode(value: <Sqlite as Database>::ValueRef<'r>) -> Result<Self, BoxDynError> {
        let state: &str = Decode::<Sqlite>::decode(value)?;
        Self::from_str(state).ok_or_else(|| -> BoxDynError {
            Box::new(InvalidClientRecordState {
                state: state.to_string(),
            })
        })
    }
}

struct SqlClientRecord {
    user_uuid: Uuid,
    user_domain: Fqdn,
    client_record_state: ClientRecordState,
    created_at: DateTime<Utc>,
    is_default: bool,
    encrypted_dek: EncryptedDek,
}

impl From<SqlClientRecord> for ClientRecord {
    fn from(value: SqlClientRecord) -> Self {
        Self {
            user_id: UserId::new(value.user_uuid, value.user_domain),
            client_record_state: value.client_record_state,
            created_at: value.created_at,
            is_default: value.is_default,
            encrypted_dek: value.encrypted_dek,
        }
    }
}

impl ClientRecord {
    pub async fn load_all_from_air_db(air_db_path: &str) -> sqlx::Result<Vec<Self>> {
        let pool = open_air_db(air_db_path).await?;
        Self::load_all(&pool).await
    }

    pub async fn load_all(executor: impl SqliteExecutor<'_>) -> sqlx::Result<Vec<Self>> {
        let records = query_as!(
            SqlClientRecord,
            r#"
            SELECT
                user_uuid AS "user_uuid: _",
                user_domain AS "user_domain: _",
                record_state AS "client_record_state: _",
                created_at AS "created_at: _",
                is_default,
                encrypted_dek AS "encrypted_dek: _"
            FROM client_record"#
        )
        .fetch_all(executor)
        .await?;
        Ok(records.into_iter().map(From::from).collect())
    }

    pub(crate) async fn load(
        executor: impl SqliteExecutor<'_>,
        user_id: &UserId,
    ) -> sqlx::Result<Option<Self>> {
        let uuid = user_id.uuid();
        let domain = user_id.domain();
        query_as!(
            SqlClientRecord,
            r#"SELECT
                user_uuid AS "user_uuid: _",
                user_domain AS "user_domain: _",
                record_state AS "client_record_state: _",
                created_at AS "created_at: _",
                is_default,
                encrypted_dek AS "encrypted_dek: _"
            FROM client_record WHERE user_uuid = ? AND user_domain = ?"#,
            uuid,
            domain
        )
        .fetch_optional(executor)
        .await
        .map(|res| res.map(From::from))
    }

    pub(super) async fn store(&self, executor: impl SqliteExecutor<'_>) -> sqlx::Result<()> {
        let record_state_str = match self.client_record_state {
            ClientRecordState::InProgress => "in_progress",
            ClientRecordState::Finished => "finished",
        };
        let uuid = self.user_id.uuid();
        let domain = self.user_id.domain();
        query!(
            "INSERT OR REPLACE INTO client_record
            (user_uuid, user_domain, record_state, created_at, is_default, encrypted_dek)
            VALUES (?1, ?2, ?3, ?4, ?5, ?6)",
            uuid,
            domain,
            record_state_str,
            self.created_at,
            self.is_default,
            self.encrypted_dek,
        )
        .execute(executor)
        .await?;
        Ok(())
    }

    pub async fn set_default(
        executor: impl SqliteExecutor<'_>,
        user_id: &UserId,
    ) -> sqlx::Result<()> {
        let uuid = user_id.uuid();
        let domain = user_id.domain();
        query!(
            "UPDATE client_record SET is_default = (user_uuid == ? AND user_domain == ?)",
            uuid,
            domain,
        )
        .execute(executor)
        .await?;
        Ok(())
    }

    pub(crate) async fn delete(
        executor: impl SqliteExecutor<'_>,
        user_id: &UserId,
    ) -> sqlx::Result<()> {
        let uuid = user_id.uuid();
        let domain = user_id.domain();
        query!(
            "DELETE FROM client_record WHERE user_uuid = ? AND user_domain = ?",
            uuid,
            domain
        )
        .execute(executor)
        .await?;
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use std::sync::LazyLock;

    use aircommon::{
        crypto::ear::{AeadCiphertext, keys::DatabaseKek},
        messages::push_token::{PushToken, PushTokenOperator},
    };
    use chrono::{DateTime, Utc};
    use sqlx::SqlitePool;
    use uuid::Uuid;

    use crate::clients::create_user::BasicUserData;

    use super::*;

    fn new_client_record(id: Uuid, created_at: DateTime<Utc>) -> ClientRecord {
        let user_id = UserId::new(id, "localhost".parse().unwrap());
        let kek = DatabaseKek::from_bytes(*b"default_db_kek______with_padding");
        let mut record = ClientRecord::new(user_id, &kek).unwrap();
        record.created_at = created_at;
        record.encrypted_dek = AeadCiphertext::dummy().into();
        record.client_record_state = ClientRecordState::Finished;
        record
    }

    fn test_client_record() -> ClientRecord {
        new_client_record(Uuid::new_v4(), Utc::now())
    }

    #[sqlx::test]
    async fn persistence(pool: SqlitePool) -> anyhow::Result<()> {
        let mut alice_record = test_client_record();
        let mut bob_record = test_client_record();

        // Storing and loading client records works
        alice_record.store(&pool).await?;
        bob_record.store(&pool).await?;
        let records = ClientRecord::load_all(&pool).await?;
        assert_eq!(records, [alice_record.clone(), bob_record.clone()]);

        // Set default to alice set alice is_default
        alice_record.is_default = true;
        ClientRecord::set_default(&pool, &alice_record.user_id).await?;
        let records = ClientRecord::load_all(&pool).await?;
        assert_eq!(records, [alice_record.clone(), bob_record.clone()]);

        // Set default to bob clears alice is_default
        alice_record.is_default = false;
        bob_record.is_default = true;
        ClientRecord::set_default(&pool, &bob_record.user_id).await?;
        let records = ClientRecord::load_all(&pool).await?;
        assert_eq!(records, [alice_record.clone(), bob_record.clone()]);

        // Delete client records
        ClientRecord::delete(&pool, &alice_record.user_id).await?;
        ClientRecord::delete(&pool, &bob_record.user_id).await?;
        let records = ClientRecord::load_all(&pool).await?;
        assert_eq!(records, []);

        Ok(())
    }

    static USER_CREATION_STATE_BASIC: LazyLock<UserCreationState> = LazyLock::new(|| {
        let user_id = Uuid::from_u128(1);

        UserCreationState::BasicUserData(BasicUserData {
            user_id: UserId::new(user_id, "localhost".parse().unwrap()),
            server_url: "localhost".to_owned(),
            push_token: Some(PushToken::new(
                PushTokenOperator::Google,
                "token".to_owned(),
            )),
        })
    });

    #[test]
    fn user_creation_state_basic_serde_codec() {
        insta::assert_binary_snapshot!(
            ".cbor",
            AirCodec::to_vec(&*USER_CREATION_STATE_BASIC).unwrap()
        );
    }

    #[test]
    fn user_creation_state_basic_json_codec() {
        insta::assert_json_snapshot!(&*USER_CREATION_STATE_BASIC);
    }

    #[test]
    fn client_record_serde_codec() {
        let record = new_client_record(Uuid::from_u128(1), "2025-01-01T00:00:00Z".parse().unwrap());
        insta::assert_binary_snapshot!(".cbor", AirCodec::to_vec(&record).unwrap());
    }

    #[test]
    fn client_record_serde_json() {
        let record = new_client_record(Uuid::from_u128(1), "2025-01-01T00:00:00Z".parse().unwrap());
        insta::assert_json_snapshot!(&record);
    }
}
