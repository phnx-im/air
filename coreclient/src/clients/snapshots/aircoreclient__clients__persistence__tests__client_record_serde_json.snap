---
source: coreclient/src/clients/persistence.rs
expression: "&record"
---
{
  "user_id": {
    "uuid": "00000000-0000-0000-0000-000000000001",
    "domain": {
      "domain": {
        "Domain": "localhost"
      }
    }
  },
  "client_record_state": "Finished",
  "created_at": "2025-01-01T00:00:00Z",
  "is_default": false,
  "encrypted_dek": {
    "ciphertext": [
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      1
    ],
    "nonce": [
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      1
    ]
  }
}
